lib/runtime/src/pipeline/network/manager.rs

```diff
@@ -19,8 +19,25 @@ use crate::distributed::RequestPlaneMode;
 use anyhow::Result;
 use async_once_cell::OnceCell;
 use std::sync::Arc;
+use std::sync::OnceLock;
 use tokio_util::sync::CancellationToken;
 
+/// Global storage for the actual TCP RPC port after binding.
+/// Uses OnceLock since the port is set once when the server binds and never changes.
+static ACTUAL_TCP_RPC_PORT: OnceLock<u16> = OnceLock::new();
+
+/// Get the actual TCP RPC port that the server is listening on.
+/// Returns 0 if the TCP server hasn't been started yet.
+pub fn get_actual_tcp_rpc_port() -> u16 {
+    ACTUAL_TCP_RPC_PORT.get().copied().unwrap_or(0)
+}
+
+/// Set the actual TCP RPC port (called internally after server binds).
+fn set_actual_tcp_rpc_port(port: u16) {
+    // Ignore if already set (shouldn't happen in normal operation)
+    let _ = ACTUAL_TCP_RPC_PORT.set(port);
+}
```

Comment:
Maybe we can return the set result, and let the caller to handle the unsuccessful case, instead of hiding it?
```suggestion
/// Set the actual TCP RPC port (called internally after server binds).
fn set_actual_tcp_rpc_port(port: u16) -> Result<(), u16> {
    ACTUAL_TCP_RPC_PORT.set(port);
}
```